### ISSUE TITLE
Provide better contextual display for patterns with large elisions

### ISSUE THREAD

**Issue by xmo-odoo** [semgrep/semgrep#3304]
**Is your feature request related to a problem? Please describe.**
Given a pattern like:
```javascript
var $JWT = require("jsonwebtoken");
...
$JWT.verify($P, $X, {algorithms:[...,'none',...]},...);
```
(taken straight from the default JS ruleset), if the `...` in the middle is very large (I've seen one extreme case where there are thousands of lines between the `require` call and the use, though a few hundreds is pretty common) semgrep will only show the initial part of the match.

**Describe the solution you'd like**
Ideally, semgrep would understand that the ellipsis (`...`) is "uninteresting" content, and would only show hunks surrounding "interesting" content. Either using `max-lines-per-finding` divided over the number of matches, or showing `max-lines-per-finding` per non-overlapping segments. Not too sure what's best there.

**Describe alternatives you've considered**
Using `--max-lines-per-finding` but that makes the output unreadable as most of it will be intermediate crap.

**Aside**
Now sure what should happen exactly if the part following the ellipsis has *multiple* matches (or what happens currently for that matter) e.g. if one import is mis-used multiple times, does semgrep-core find just the first as a single match, all of them as a single match, or does it treat each as different matches (and the deduplication is done afterwards)?

The CLI only shows a single match but I don't know if that's what the core generates or something the CLI does afterwards.

**Comment by aryx** [comment]
That's a great idea!

**Comment by xmo-odoo** [comment]
Stalebot is the stupidest bot on this damn platform.

**Comment by IagoAbal** [comment]
In the meantime, would it work to rewrite the pattern this way?
```
- patterns:
  - pattern-inside: |
       var $JWT = require("jsonwebtoken");
       ...
  - pattern: |
       $JWT.verify($P, $X, {algorithms:[...,'none',...]},...);
```

Letting Semgrep CLI know what part of the match corresponds to `...` sounds nice, but I think it's not a quick thing to do. Perhaps Semgrep CLI could look at the lines where the metavariables match, and mark those lines as "important" to make sure that they are always in the output match.

**Comment by xmo-odoo** [comment]
@IagoAbal thanks for the suggestion, I really need to get more comfortable with `pattern-inside`, I probably don't reach for it anywhere near enough.

**Comment by aryx** [comment]
:wave:

**Comment by xmo-odoo** [comment]
![dogconfused](https://user-images.githubusercontent.com/7571158/150346355-44aacc38-2b2f-46f1-b0a2-9b5e5516bd57.png)

**Comment by xmo-odoo** [comment]
![animal 2](https://user-images.githubusercontent.com/7571158/152739439-049a798e-847d-41a7-a3a1-38de390aaf72.png)

**Comment by IagoAbal** [comment]
We need to set some priority here. cc @ievans 

> Letting Semgrep CLI know what part of the match corresponds to ... sounds nice, but I think it's not a quick thing to do. Perhaps Semgrep CLI could look at the lines where the metavariables match, and mark those lines as "important" to make sure that they are always in the output match.

I still think that the easiest solution would be for the CLI to make sure that the lines where metavariables match are included in the output. That is not perfect but it should do the job, and it doesn't require any change in the matching engine. CCing @brendongo who is the CLI expert. The CLI could also apply some other heuristics, if needed.

I am not sure if we allow `$...MVAR` in every place where `...` is allowed. If we did, then a workaround may be to use some `$...INTERNAL_ELLIPSIS_XYZ`. Then the CLI could know what `...` is matching. But doing that would likely make matching `...` more expensive since we will have to record the matched code (thus adding more allocations). cc @aryx

**Comment by ievans** [comment]
@xmo-odoo we have banished the stale bot from this issue :P 

Have you looked at https://github.com/returntocorp/semgrep/pull/4735? It's not documented yet, but does it address this use-case too?

**Comment by xmo-odoo** [comment]
> we have banished the stale bot from this issue :P

Yay.

> Have you looked at #4735? It's not documented yet, but does it address this use-case too?

It's not entirely clear what it does but I don't think so? Reading the issue it looks like a shortcut for `metavariable-pattern`, something along those lines? So an improvement in matching semantics?

But here the issue is not one of matching semantics, the matching semantics are completely fine, the issue is one of display rendering: when matching a pattern with a large amount of "dead space" (the start and end of the pattern are far apart in the code), semgrep is currently unable to show the "source and sink" and will only show the start of the match, the user has to move to e.g. an editor for further analysis / insight into the callsite.

@IagoAbal's example using `pattern-inside` and `pattern` works around the issue by basically dropping the start of the pattern (the source) using `pattern-inside`, and only showing the sink. TBF it's usually the most important bit, and thus I think will work in most cases, but there are others where the source is quite relevant. 

I (still) think in a `pattern` the ellipsis tend to be for "uninteresting" content, and thus can usually be elided in order to provide a better rendering of the "edges" of the match.

**Comment by IagoAbal** [comment]
I'll be adding some docs for #4735 soon but it somewhat relates to this, as it allows to "zoom in", and "focus" a match in a particular region of interest (bound by a metavariable).

I've been thinking about this and implementing it may not be difficult, I just don't know how it may affect performance. We should try. But also note that if you use taint-mode then Semgrep knows already what is a source and what is a sink, and it would be easy to report findings in a better way (and I believe we will be doing this soon-ish).

Another option would be allowing for a richer `message:` perhaps? If the "source" is important, can't we just add it to the rule message?

**Comment by xmo-odoo** [comment]
> I've been thinking about this and implementing it may not be difficult, I just don't know how it may affect performance. We should try. But also note that if you use taint-mode then Semgrep knows already what is a source and what is a sink, and it would be easy to report findings in a better way (and I believe we will be doing this soon-ish).

Yeah that's pretty much why I used the terminology, I figured either this already existed for traint tracking (and it might be reusable) or it doesn't (and it might be useful there as well to explain the "taint flow").

> Another option would be allowing for a richer `message:` perhaps? If the "source" is important, can't we just add it to the rule message?

That would mostly be for the same situation as the workaround you suggested where the start of the pattern is mostly static yeah? Or do you mean messages support metavariable expansion so the "dynamic" bits of the pattern start could be injected there and thus fully replicate said pattern?

**Comment by IagoAbal** [comment]
> Or do you mean messages support metavariable expansion so the "dynamic" bits of the pattern start could be injected there and thus fully replicate said pattern?

Yes, that's kind of what I mean. Messages already support metavariable expansion, but it has some limitations. I think e.g. you cannot print source code locations. We could make it "richer" in different ways, e.g. allowing to easily add a particular line of code (line info included). I find that could be a more general solution.

The only problem is that we _may_ need to extend the pattern syntax to capture the relevant pieces of code we want to use in the `message:` field, e.g. something like:

```js
- pattern-inside: |
    var $JWT = require("jsonwebtoken");[@$SOURCE]
    ...
- pattern: |
    $JWT.verify($P, $X, {algorithms:[...,'none',...]},...);
```

where `[@$SOURCE]` would make Semgrep capture that statement in the metavariable `$SOURCE`, which then you could use in `message:`. Although maybe we can find some simpler way.

**Comment by xmo-odoo** [comment]
> Yes, that's kind of what I mean. Messages already support metavariable expansion

Oh TIL, I had no idea. That's good to know, and it does probably resolve most of the issue then (as long as one remembers to use `pattern-inside`), at least wrt reporting. Even if parts of the leading edge have to be copied into the message that's at least reasonably feasible.
